service/history/workflow/update/registry.go

```diff
@@ -224,14 +233,21 @@ func (r *RegistryImpl) remover(id string) updateOpt {
 	)
 }
 
-func (r *RegistryImpl) admit(context.Context) error {
-	max := r.maxInFlight()
-	if len(r.updates) >= max {
+func (r *RegistryImpl) admit(ctx context.Context) error {
+	if len(r.updates) >= r.maxInFlight() {
+		return serviceerror.NewResourceExhausted(
+			enumspb.RESOURCE_EXHAUSTED_CAUSE_CONCURRENT_LIMIT,
+			fmt.Sprintf("update concurrent in-flight limit has been reached (%v)", r.maxInFlight()),
+		)
+	}
+
+	if r.store.GetUpdatesCount(ctx) >= r.maxTotal() {
 		return serviceerror.NewResourceExhausted(
 			enumspb.RESOURCE_EXHAUSTED_CAUSE_CONCURRENT_LIMIT,
```

Comment:
max total update is not concurrent limit? InvalidArgument or FailedPrecondition?

Reply:
Yes, `FailedPrecondtion` is a better code here.